PhysicsTools/PatAlgos/plugins/PATTracksToPackedCandidates.cc

```diff
@@ -0,0 +1,203 @@
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+#include "DataFormats/Common/interface/Association.h"
+
+#include <vector>
+
+//
+// class declaration
+//
+
+class PATTracksToPackedCandidates : public edm::stream::EDProducer<> {
+public:
+  explicit PATTracksToPackedCandidates(const edm::ParameterSet&);
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  void addPackedCandidate(std::vector<pat::PackedCandidate>& cands,
+                          const reco::Track trk,
+                          const reco::VertexRef& pvSlimmed,
+                          const reco::VertexRefProd& pvSlimmedColl,
+                          bool passPixelTrackSel) const;
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  // ----------member data ---------------------------
+  const edm::EDGetTokenT<reco::TrackCollection> srcTracks_;
+  const edm::EDGetTokenT<reco::VertexCollection> srcPrimaryVertices_;
+  const double dzSigCut_;
+  const double dxySigCut_;
+  const double dzSigHP_;
+  const double dxySigHP_;
+  const double ptMax_;
+  const double ptMin_;
+  const bool resetHP_;
+  const int covarianceVersion_;
+  const int covarianceSchema_;
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+PATTracksToPackedCandidates::PATTracksToPackedCandidates(const edm::ParameterSet& iConfig)
+    : srcTracks_(consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("srcTracks"))),
+      srcPrimaryVertices_(consumes<reco::VertexCollection>(iConfig.getParameter<edm::InputTag>("srcPrimaryVertices"))),
+      dzSigCut_(iConfig.getParameter<double>("dzSigCut")),
+      dxySigCut_(iConfig.getParameter<double>("dxySigCut")),
+      dzSigHP_(iConfig.getParameter<double>("dzSigHP")),
+      dxySigHP_(iConfig.getParameter<double>("dxySigHP")),
+      ptMax_(iConfig.getParameter<double>("ptMax")),
+      ptMin_(iConfig.getParameter<double>("ptMin")),
+      resetHP_(iConfig.getParameter<bool>("resetHP")),
+      covarianceVersion_(iConfig.getParameter<int>("covarianceVersion")),
+      covarianceSchema_(iConfig.getParameter<int>("covarianceSchema")) {
+  produces<std::vector<pat::PackedCandidate>>();
+  produces<edm::Association<pat::PackedCandidateCollection>>();
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void PATTracksToPackedCandidates::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  using namespace std;
+
+  //track collection
+  auto tracks = iEvent.getHandle(srcTracks_);
+
+  auto outPtrTrksAsCands = std::make_unique<std::vector<pat::PackedCandidate>>();
+
+  //vtx collection
+  edm::Handle<reco::VertexCollection> pvs;
+  iEvent.getByToken(srcPrimaryVertices_, pvs);
+  reco::VertexRef pv(pvs.id());
+  reco::VertexRefProd pvRefProd(pvs);
+  if (!pvs->empty()) {
+    pv = reco::VertexRef(pvs, 0);
+  }
```

Comment:
from L113 it looks like pvs->at(0) is used without any checks for empty.
Either there should be a protection there to default to some value (like beam spot),
or the checks for empty can be removed here and the ```pv``` ref itself can be removed, because it is only used to pass its index 0 through the ```addPackedCandidate```

Reply:
I added a protection using BS, please, see commit: https://github.com/cms-sw/cmssw/pull/30313/commits/77cb1baaa80d6d11618e631ff10c2582072bf804